### PR TITLE
daemon: change when restored endpoints are inserted into endpoint manager

### LINKED ISSUE #4390: race condition with starting of L7 Proxy and regenerating endpoint
In #4329, the following build failed: https://jenkins.cilium.io/job/cilium-ginkgo/job/cilium/job/master/974/testReport/junit/k8s-1/7/K8sValidatedUpdates_Updating_Cilium_stable_to_master/

Logs: [f0c5341e_K8sValidatedUpdates_Updating_Cilium_stable_to_master.2.zip](https://github.com/cilium/cilium/files/2078265/f0c5341e_K8sValidatedUpdates_Updating_Cilium_stable_to_master.2.zip)

The pod `cilium-j5tqp` was not marked as 'ready' by the output of `kubectl get pods`:

```
cmd: kubectl get pods -o wide --all-namespaces
NAMESPACE     NAME                           READY     STATUS    RESTARTS   AGE       IP              NODE
default       app1-3318830813-4c3cq          1/1       Running   0          9m        10.10.0.32      k8s1
default       app1-3318830813-hv6zr          1/1       Running   0          9m        10.10.0.66      k8s1
default       app2-3691078314-qrdzw          1/1       Running   0          9m        10.10.0.59      k8s1
default       app3-753781860-8dzzn           1/1       Running   0          9m        10.10.0.225     k8s1
kube-system   cilium-5xm73                   1/1       Running   0          5m        192.168.36.12   k8s2
kube-system   cilium-j5tqp                   0/1       Running   1          5m        192.168.36.11   k8s1
kube-system   etcd-k8s1                      1/1       Running   0          47m       192.168.36.11   k8s1
kube-system   kube-apiserver-k8s1            1/1       Running   0          47m       192.168.36.11   k8s1
kube-system   kube-controller-manager-k8s1   1/1       Running   0          47m       192.168.36.11   k8s1
kube-system   kube-dns-1992960195-w0vp9      3/3       Running   0          47m       10.10.1.27      k8s2
kube-system   kube-proxy-d2k7g               1/1       Running   0          19m       192.168.36.12   k8s2
kube-system   kube-proxy-ng786               1/1       Running   0          47m       192.168.36.11   k8s1
kube-system   kube-scheduler-k8s1            1/1       Running   0          47m       192.168.36.11   k8s1
```

When trying to debug #4329 , I noticed that the logs had endpoints regenerating before the call to restore them.

In the attached archive, the logs for `cilium-j5tqp` had the following ordering of logs:

```
2018-06-06T15:57:56.923139812Z level=info msg="Restoring endpoints from former life..."
...
2018-06-06T15:57:56.988674269Z level=info msg="Initializing identity allocator"
...
2018-06-06T15:57:57.000360223Z level=debug msg="Forced policy recalculation triggered"
2018-06-06T15:57:57.00058318Z level=debug msg="Forced policy recalculation triggered"
2018-06-06T15:57:57.000796957Z level=debug msg="Forced policy recalculation triggered"
2018-06-06T15:57:57.001006206Z level=debug msg="Forced policy recalculation triggered"
2018-06-06T15:57:57.001319466Z level=debug msg="Forced policy recalculation triggered"
...
2018-06-06T15:57:57.004242003Z level=debug msg="Regenerating endpoint..." containerID=08cd77e1b4 endpointID=29381 ipv4=10.10.0.59 ipv6="f00d::a0a:0:0:72c5" k8sPodName=":" policyRevision=0
...
2018-06-06T15:57:57.050352649Z level=warning msg="Regeneration of endpoint program failed" containerID=bd1258b3a6 endpointID=25729 error="can't update network policy, proxy disabled" ipv4=10.10.0.225 ipv6="f00d::a0a:0:0:6481" k8sPodName=":" policyRevision=0
2018-06-06T15:57:57.050703775Z level=debug msg="Error regenerating endpoint: can't update network policy, proxy disabled" code=Failure containerID=bd1258b3a6 endpointID=25729 endpointState=ready ipv4=10.10.0.225 ipv6="f00d::a0a:0:0:6481" k8sPodName=":" policyRevision=0 type=200
...
2018-06-06T15:58:02.211073997Z level=info msg="Regenerating 4 restored endpoints"
...
2018-06-06T15:58:02.211122952Z level=info msg="Starting IP identity watcher"
...
2018-06-06T15:58:19.219401821Z level=info msg="Finished regenerating restored endpoints" regenerated=4 total=4
...
2018-06-06T15:58:19.416073671Z level=info msg="Envoy: Accepted access log connection"
```

The logs above indicate that the regeneration of endpoint BPF programs / configuration of the proxy is done before the proxy is done in `daemon/daemon.go:NewDaemon`:

```
restoredEndpoints, err := d.restoreOldEndpoints(option.Config.StateDir, true)
...
identity.InitIdentityAllocator(&d)
...
d.l7Proxy = proxy.StartProxySupport(10000, 20000, option.Config.RunDir, option.Config.AccessLog, &d, option.Config.AgentLabels)

if option.Config.RestoreState {
    d.regenerateRestoredEndpoints(restoredEndpoints)
...
```

In the call to `d.restoreOldEndpoints`, the endpoint is not regenerated yet, but is inserted into the endpoint manager:

https://github.com/cilium/cilium/blob/91287b9f0665ddd22fa4933eece6ae276e4157af/daemon/state.go#L113

And then, in the call to `InitIdentityAllocator`, whenever an Identity is received from the key-value store, we forcibly trigger policy updates:

https://github.com/cilium/cilium/blob/91287b9f0665ddd22fa4933eece6ae276e4157af/pkg/identity/cache.go#L66-L78

This in turn triggers regeneration of the endpoints, which can occur before the `l7Proxy` for `daemon` is initialized. This results in endpoints failing to regenerate upon restart of the agent. Luckily, in the call to `d.regenerateRestoredEndpoints`, the endpoints are able to be put into an OK state, but this means that we are trying to regenerate endpoints before all of the initialization of various subsystems in `cilium-agent` is complete.

This will impact running endpoints and their BPF programs because the endpoints will be put into `not-ready` state, as the error with the `l7Proxy` being nil is hit in the code path started here:

https://github.com/cilium/cilium/blob/91287b9f0665ddd22fa4933eece6ae276e4157af/pkg/endpoint/bpf.go#L687-L691

This will result in the destruction of the BPF PolicyMap / CIDR policy maps for the endpoints, which may cause a loss of packets for existing workloads.

### PR DESCRIPTION
**Pull request by ianvernon**
`daemon/state.go:restoreOldEndpoints` is executed before
`pkg/identity/allocator.go.InitIdentityAllocator`. Previously, endpoints were inserted into the endpointmanager in `d.restoreOldEndpoints`.
`pkg/identity/allocator.go.InitIdentityAllocator` calls `daemon/policy.go:TriggerPolicyUpdates`, which causes regeneration of policy for all endpoints managed by the endpointmanager. `pkg/identity/allocator.go.InitIdentityAllocator` is called before the `daemon.l7Proxy` is bootstrapped. This would result in errors like the following:

```
Error regenerating endpoint: can't update network policy, proxy disabled
```

which can result in the endpoints being put into `not-ready` state. This is because endpoints were inserted into the endpointmanager before `daemon.l7Proxy` was bootstrapped.

To fix this, move the insertion of endpoints into the endpoint manager in `daemon/state.go:regenerateRestoredEndpoints`, which is called after `daemon.l7Proxy`is bootstrapped, and after `pkg/identity/allocator.go.InitIdentityAllocator`. Since it is not desirable to try to regenerate endpoints until `daemon.l7Proxy` is bootstrapped, and `daemon/state.go:regenerateRestoredEndpoints` regenerates the endpoints, it is guaranteed that the above error will not be hit because `daemon.l7Proxy` is created synchronously before `daemon/state.go:regenerateRestoredEndpoints` is called.

Signed-off-by: Ian Vernon <ian@cilium.io>

**Summary of changes**:

Fixes: #4390

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4398)

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

File: daemon/state.go [hunk 1]
```diff
@@ -134,6 +133,11 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState) {
...
 		go func(ep *endpoint.Endpoint, epRegenerated chan<- bool) {
 			ep.Mutex.Lock()
+
+			// Insert into endpoint manager so it can be regenerated when calls to
+			// TriggerPolicyUpdates() are made.
+			endpointmanager.Insert(ep)
+
 			scopedLog := log.WithField(logfields.EndpointID, ep.ID)
 
 			state := ep.GetStateLocked()
```
Comment (tgraf) [inline daemon/state.go]: Given this change, you can remove this code do detect and abort on disconnected state again as there is no way that an endpoint could have been disconnected in the meantime as it was not accessible in any way.

**Review by tgraf — DISMISSED** [review]

**Comment by raybejjani** [comment]
This is probably fine but endpointmanager.Insert also updates some lookups for each EP. Those are used in the workload watchers to correlate events (I think). Since we've delayed the inserts to after all the restores are complete, and to after `InitIdentityAllocator` and `daemon.go:init` this might make our startup behaviour a little weird because of the large delay. In particular, I would expect that we might ignore containers for a little longer, until we do a periodic full sync of that data.

File: daemon/state.go [hunk 2]
```diff
@@ -134,6 +133,11 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState) {
 	for _, ep := range state.restored {
 		go func(ep *endpoint.Endpoint, epRegenerated chan<- bool) {
 			ep.Mutex.Lock()
+
+			// Insert into endpoint manager so it can be regenerated when calls to
+			// TriggerPolicyUpdates() are made.
+			endpointmanager.Insert(ep)
```
Comment (aanm) [inline daemon/state.go]: The only issue I see with this is the `cilium endpoint list` can be empty as soon the `Endpoints restored, endpoints will be regenerated in background` is printed and the endpoints will show up afterwards. Might confuse users, no?

**Review by aanm — DISMISSED** [review]

File: daemon/state.go (hunk 2 above)
Comment (ianvernon) [reply daemon/state.go]: I'll remove the 'regenerated in background' part of the message.

**Comment by ianvernon** [comment]
> This is probably fine but endpointmanager.Insert also updates some lookups for each EP. Those are used in the workload watchers to correlate events (I think). Since we've delayed the inserts to after all the restores are complete, and to after InitIdentityAllocator and daemon.go:init this might make our startup behaviour a little weird because of the large delay. In particular, I would expect that we might ignore containers for a little longer, until we do a periodic full sync of that data.

I'm unsure how large this delay will be - especially since there will be no endpoints to regenerate when `InitIdentityAllocator` is initially called. We can try shuffling around where we initialize certain parts of the subsystem in follow-up PRs to optimize maybe? :/

**Comment by ianvernon** [comment]
test-me-please

**Comment by joestringer** [comment]
@raybejjani @ianvernon wasn't the entire restore previously happening around the place where this PR is proposing to put the `endpointmanager.Insert()`? We just moved parts of that workflow earlier to fix races with the IPAM?

**Comment by ianvernon** [comment]
@joestringer it appears so, yes.

**Review by tgraf — APPROVED** [review]

File: daemon/state.go (hunk 2 above)
Comment (ianvernon) [reply daemon/state.go]: Discussed and resolved offline.

File: daemon/state.go [hunk 3]
```diff
@@ -134,14 +133,12 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState) {
 	for _, ep := range state.restored {
 		go func(ep *endpoint.Endpoint, epRegenerated chan<- bool) {
 			ep.Mutex.Lock()
-			scopedLog := log.WithField(logfields.EndpointID, ep.ID)
 
-			state := ep.GetStateLocked()
-			if state == endpoint.StateDisconnected || state == endpoint.StateDisconnecting {
```
Comment (manalibhutiyani) [inline daemon/state.go:140]: Why did we need to delete this logic? 
wouldn't this still be true , i.e can we not receive endpoints in regenerateRestoredEndpoints in state
`endpoint.StateDisconnected` or `state == endpoint.StateDisconnecting`?

Comment (ianvernon) [reply daemon/state.go:140]: See https://github.com/cilium/cilium/pull/4398#discussion_r193624510

**Pull request merged**